src/main/java/com/hotels/beeju/HiveServer2JUnitRule.java

```diff
@@ -0,0 +1,191 @@
+
+import com.hotels.beeju.server2.RelaxedSQLStdHiveAuthorizerFactory;
+
+/**
+ * A JUnit Rule that creates a Hive Server 2 service and Thrift service backed by a Hive Metastore using an HSQLDB
+ * in-memory database.
+ * <p>
+ * A fresh database instance will be created for each test method.
+ * <p>
+ * Note that this class is meant to be used for DDL operations only. Any attempt to run DML operations may results in
+ * errors.
+ */
+public class HiveServer2JUnitRule extends ExternalResource {
+
+  public static final String HSQLDB_USER = "user";
+  public static final String HSQLDB_PASSWORD = "password";
+
+  @VisibleForTesting
+  final TemporaryFolder temporaryFolder = new TemporaryFolder();
+  protected final HiveConf conf = new HiveConf();
+  private final String databaseName;
+  private final String metastoreConnectionURL;
+  private final String driverClassName;
+  private String connectionUrl;
+  private HiveServer2 hiveServer2;
+  private File metastoreLocation;
+  private int port;
+
+  /**
+   * Create a Hive Server 2 service with a pre-created database "test_database".
+   */
+  public HiveServer2JUnitRule() {
+    this("test_database");
+  }
+
+  /**
+   * Create a Hive Server 2 service with a pre-created database using the provided name.
+   *
+   * @param databaseName Database name.
+   */
+  public HiveServer2JUnitRule(String databaseName) {
+    checkNotNull(databaseName, "databaseName is required");
+    this.databaseName = databaseName;
+    driverClassName = JDBCDriver.class.getName();
+    conf.setBoolean("hcatalog.hive.client.cache.disabled", true);
+    metastoreConnectionURL = "jdbc:hsqldb:mem:" + UUID.randomUUID();
+    conf.setVar(ConfVars.METASTORECONNECTURLKEY, metastoreConnectionURL);
+    conf.setVar(ConfVars.METASTORE_CONNECTION_DRIVER, driverClassName);
+    conf.setVar(ConfVars.METASTORE_CONNECTION_USER_NAME, HSQLDB_USER);
+    conf.setVar(ConfVars.METASTOREPWD, HSQLDB_PASSWORD);
+    conf.setBoolVar(ConfVars.HMSHANDLERFORCERELOADCONF, true);
+    conf.setVar(ConfVars.HIVE_AUTHORIZATION_MANAGER, "");
+    // Hive 2.x compatibility
+    conf.setBoolean("datanucleus.schema.autoCreateAll", true);
+  }
+
+  @Override
+  protected void before() throws Throwable {
+    temporaryFolder.create();
+    init();
+    startHiveServer2();
+  }
+
+  private void init() throws IOException {
+    metastoreLocation = temporaryFolder.newFolder("metastore");
+    conf.setVar(ConfVars.METASTOREWAREHOUSE, metastoreLocation.getAbsolutePath());
+    try (ServerSocket s = new ServerSocket(0)) {
+      port = s.getLocalPort();
+    }
+    conf.setIntVar(HiveConf.ConfVars.HIVE_SERVER2_THRIFT_PORT, port);
+  }
+
+  private void startHiveServer2() throws Throwable {
+    conf.setVar(HiveConf.ConfVars.HIVE_AUTHORIZATION_MANAGER, RelaxedSQLStdHiveAuthorizerFactory.class.getName());
+    hiveServer2 = new HiveServer2();
+    hiveServer2.init(conf);
+    hiveServer2.start();
+    Thread.sleep(5000);
```

Comment:
We can do better than them. How about replacing the Thread.sleep(5000) with:

`
    int retries = 0;
    int maxRetries = 5;
    while (hiveServer2.getServiceState() != STATE.STARTED && retries < maxRetries) {
      Thread.sleep(1000);
      retries++;
    }
    if (retries > maxRetries) {
      throw new RuntimeException("HiveServer2 did not start in a reasonable time");
    }
`

(I checked with this code and all the tests pass)